securetar/__init__.py

```diff
@@ -702,6 +677,149 @@ def _finalize_tar_entry(self) -> None:
+            raise ValueError("Either filename or fileobj must be provided")
+
+        if mode not in (MOD_EXCLUSIVE, MOD_READ, MOD_WRITE):
+            raise ValueError(
+                f"Mode must be '{MOD_EXCLUSIVE}', '{MOD_READ}', or '{MOD_WRITE}'"
+            )
+
+        self._name = name
+        self._mode = mode
+        self._bufsize = bufsize
+        self._fileobj = fileobj
+        self._password = password
+        self._streaming = streaming
+        self._tar: tarfile.TarFile | None = None
+
+    def __enter__(self) -> tarfile.TarFile:
+        """Open the archive."""
+        return self.open()
+
+    def open(self) -> tarfile.TarFile:
+        """Open the archive."""
+        mode=f"{self._mode}{'|' if self._streaming else ''}"
+        self._tar = tarfile.open(
+            name=str(self._name) if self._name else None,
+            mode=mode,
+            fileobj=self._fileobj,
+            bufsize=self._bufsize,
+        )
+        return self._tar
+
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        """Close the archive."""
+        self.close()
+
+    def close(self) -> None:
+        """Close the archive."""
+        if self._tar:
+            self._tar.close()
+            self._tar = None
+
+    def create_inner_tar(
+        self,
+        name: str,
+        *,
+        gzip: bool = True,
+    ) -> _InnerSecureTarFile:
+        """Create an inner tar file, optionally encrypted.
+
+        Args:
+            name: Name of the inner tar file in the archive
+            gzip: Whether to use gzip compression
+        """
+        if not self._tar:
+            raise SecureTarError("Archive not open")
+        if self._streaming:
+            raise SecureTarError("create_inner_tar not supported in streaming mode")
+
+        return _InnerSecureTarFile(
+            self._tar,
+            bufsize=self._bufsize,
+            gzip=gzip,
+            mode="w",
+            name=Path(name),
+            password=self._password,
+        )
+
+    def open_inner_tar(
+        self,
+        member: tarfile.TarInfo,
+    ) -> _SecureTarDecryptingStream:
+        """Open an encrypted inner tar file for reading.
+
+        Returns a context manager that yields a readable stream of decrypted bytes.
+        The stream can be passed to tarfile.open() to read the inner tar contents.
+        """
+        if not self._tar:
+            raise SecureTarError("Archive not open")
+
+        if self._mode != MOD_READ:
+            raise SecureTarError("Archive not open for reading")
+
+        if not self._password:
+            raise SecureTarError("No password provided")
+
+        fileobj = self._tar.extractfile(member)
+        if fileobj is None:
+            raise SecureTarError(f"Cannot extract {member.name}")
+
+        return _SecureTarDecryptingStream(
+            fileobj,
+            self._password,
+            ciphertext_size=member.size,
+        )
+
+    def add_inner_tar(
```

Comment:
It's not clear to me what the difference between `create_inner_tar` and `add_inner_tar` is. Do you have to create an inner tar file before you can use `add_inner_tar`?

Reply:
I agree the naming is not great, but your understanding is correct.

This is a summary of the methods:
- `create_inner_tar`: Create an empty stdlib `tarfile.TarFile` and add it to the outer archive. The new TarFile will be opened in write mode and the user can add files to it. This is supported both with and without encryption.
- `add_inner_tar`: Add an already prepared, non encrypted, tar to the outer archive. Only supported with encryption.
- `open_inner_tar`: Allows reading a tar from the outer archive. Only supported with encryption.